cdc/sink/codec/maxwell.go

```diff
@@ -0,0 +1,470 @@
+		return 0, false, err
+	}
+	return b.nextKey.Type, true, nil
+}
+
+// decode
+func maxwellMessageToRowEvent(key *mqMessageKey, value *maxwellMessage) *model.RowChangedEvent {
+	e := new(model.RowChangedEvent)
+	e.CommitTs = value.Ts
+	e.Table = &model.TableName{
+		Schema: key.Schema,
+		Table:  key.Table,
+	}
+	if key.Partition != nil {
+		e.Table.TableID = *key.Partition
+		e.Table.IsPartition = true
+	}
+	if value.Old == nil {
+		for k, v := range value.Data {
+			for _, n := range e.Columns {
+				n.Name = k
+				n.Value = v
+				n.Type = 3
+
+			}
+		}
+	}
+
+	return e
+}
+
+// NextResolvedEvent implements the EventBatchDecoder interface
+func (b *MaxwellEventBatchDecoder) NextResolvedEvent() (uint64, error) {
+	// For maxwell now, there is no such a corresponding type to ResolvedEvent so far.
+	// Therefore the event is ignored.
+	return 0, nil
+}
+
+// NextRowChangedEvent implements the EventBatchDecoder interface
+func (b *MaxwellEventBatchDecoder) NextRowChangedEvent() (*model.RowChangedEvent, error) {
+	if b.nextKey == nil {
+		if err := b.decodeNextKey(); err != nil {
+			return nil, err
+		}
+	}
+	b.keyBytes = b.keyBytes[b.nextKeyLen+8:]
+	if b.nextKey.Type != model.MqMessageTypeRow {
+		return nil, errors.NotFoundf("not found row event message")
+	}
+	valueLen := binary.BigEndian.Uint64(b.valueBytes[:8])
+	value := b.valueBytes[8 : valueLen+8]
+	b.valueBytes = b.valueBytes[valueLen+8:]
+	rowMsg := new(maxwellMessage)
+	if err := rowMsg.Decode(value); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	rowEvent := maxwellMessageToRowEvent(b.nextKey, rowMsg)
+	b.nextKey = nil
+	return rowEvent, nil
+}
+
+// NextDDLEvent implements the EventBatchDecoder interface
+func (b *MaxwellEventBatchDecoder) NextDDLEvent() (*model.DDLEvent, error) {
+	if b.nextKey == nil {
+		if err := b.decodeNextKey(); err != nil {
+			return nil, err
+		}
+	}
+	b.keyBytes = b.keyBytes[b.nextKeyLen+8:]
+	if b.nextKey.Type != model.MqMessageTypeDDL {
+		return nil, errors.NotFoundf("not found ddl event message")
+	}
+	valueLen := binary.BigEndian.Uint64(b.valueBytes[:8])
+	value := b.valueBytes[8 : valueLen+8]
+	b.valueBytes = b.valueBytes[valueLen+8:]
+	ddlMsg := new(mqMessageDDL)
+	if err := ddlMsg.Decode(value); err != nil {
+		return nil, errors.Trace(err)
+	}
+	ddlEvent := mqMessageToDDLEvent(b.nextKey, ddlMsg)
+	b.nextKey = nil
+	return ddlEvent, nil
+}
+
+// NewMaxwellEventBatchDecoder creates a new JSONEventBatchDecoder.
+func NewMaxwellEventBatchDecoder(key []byte, value []byte) (EventBatchDecoder, error) {
+	version := binary.BigEndian.Uint64(key[:8])
+	key = key[8:]
+	if version != BatchVersion1 {
+		return nil, errors.New("unexpected key format version")
+	}
+	return &MaxwellEventBatchDecoder{
+		keyBytes:   key,
+		valueBytes: value,
+	}, nil
+}
+
+//ddl typecode from parser/model/ddl.go
+func ddlToMaxwellType(ddlType model2.ActionType) string {
+	if ddlType >= 5 && ddlType <= 20 {
+		return "table-alter"
+	}
+	switch ddlType {
+	case 3:
```

Comment:
Is it better to use constant variable from parser lib instead of the const number, such as `model2.ActionCreateTable`?

Reply:
done